### PR TITLE
Ensure that resources not specified by --target do not get added to checkpoint file.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/pulumi/pulumi/issues/3790

### CODE REVIEW

File: pkg/backend/snapshot.go [hunk 1]
```diff
@@ -249,18 +242,30 @@ func (ssm *sameSnapshotMutation) End(step deploy.Step, successful bool) error {
...
+
+		// In the case of a 'resource create' in a program that wasn't specified by the user in the
+		// --target list, we *never* want to write this to the checkpoint.  We treat it as if it
+		// doesn't exist at all.  That way when the program runs the next time, we'll actually
+		// create it.
+		if sameStep.IsSkippedCreate() {
+			return false
+		}
+
 		ssm.manager.markNew(step.New())
```
Comment (CyrusNajmabadi) [inline pkg/backend/snapshot.go]: the `markNew` was the line that was ultimately causing problems.  Even though we didn't think we needed to write this information to the snapshot (below), we still added the 'create' to the 'news' list which ended up still writing later on

File: tests/integration/targets/targets_test.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+	if err := fsutil.CopyFile(
+		path.Join(e.RootPath, "untargeted_create", "index.ts"),
+		path.Join("untargeted_create", "step1", "index.ts"), nil); err != nil {
+
+		t.Fatalf("error copying index.ts file: %v", err)
+	}
+
+	e.RunCommand("pulumi", "up", "--target", urn, "--non-interactive", "--skip-preview", "--yes")
+	e.RunCommand("pulumi refresh")
```
Comment (CyrusNajmabadi) [inline tests/integration/targets/targets_test.go]: validated this sequence of steps failed without fix and works with fix.  manually checked snapshot as well to ensure no vestigates of the created resource were in it.

**Comment by pgavlin** [comment]
Just to refresh my memory, we do require that creates are present in the target list if they are dependencies of other resources that are being targeted, yes?

**Comment by CyrusNajmabadi** [comment]
> Just to refresh my memory, we do require that creates are present in the target list if they are dependencies of other resources that are being targeted, yes?

Correct, and we have test in that area (like `TestCreateDuringTargetedUpdate_UntargetedCreateReferencedByTarget`).

**Comment by pgavlin** [comment]
Okay, great. In that case, I do believe that this should work. I was worried that there might be some way for a reference to the skipped create to sneak into the checkpoint, which would make it invalid if the resource was not written. I think that is not possible, though:
- if the skipped resource is referenced by a non-targeted resource, that resource's state (incl. dependencies) will not change by definition
- if the skipped resource is referenced by a targeted resource, the update will fail

File: pkg/backend/snapshot.go [hunk 3]
```diff
@@ -249,18 +242,30 @@ func (ssm *sameSnapshotMutation) End(step deploy.Step, successful bool) error {
...
 	return ssm.manager.mutate(func() bool {
+		sameStep := step.(*deploy.SameStep)
+
 		ssm.manager.markDone(step.Old())
+
+		// In the case of a 'resource create' in a program that wasn't specified by the user in the
+		// --target list, we *never* want to write this to the checkpoint.  We treat it as if it
+		// doesn't exist at all.  That way when the program runs the next time, we'll actually
+		// create it.
+		if sameStep.IsSkippedCreate() {
```
Comment (pgavlin) [inline pkg/backend/snapshot.go]:
```suggestion
		if sameStep := step.SameStep(); sameStep.IsSkippedCreate() {
```

And remote the assignment at the beginning of the function.

Comment (CyrusNajmabadi) [reply pkg/backend/snapshot.go]: so this only makes sameStep visible in the body of the 'if'.  but we want to reference it below as we pass it to `.mustWrite`.

Comment (pgavlin) [reply pkg/backend/snapshot.go]: Ah, missed that bit. 👍

**Pull request merged**